pkg/plugins/models.go

```diff
@@ -42,9 +43,100 @@ func (e DuplicateError) Is(err error) bool {
 }
 
 type Dependencies struct {
-	GrafanaDependency string       `json:"grafanaDependency"`
-	GrafanaVersion    string       `json:"grafanaVersion"`
-	Plugins           []Dependency `json:"plugins"`
+	GrafanaDependency string                 `json:"grafanaDependency"`
+	GrafanaVersion    string                 `json:"grafanaVersion"`
+	Plugins           []Dependency           `json:"plugins"`
+	Extensions        ExtensionsDependencies `json:"extensions"`
+}
+
+// We need different versions for the Extensions struct because there is a now deprecated plugin.json schema out there, where the "extensions" prop
+// is in a different format (Extensions V1). In order to support those as well while reading the plugin.json, we need to add a custom unmarshaling logic for extensions.
+type ExtensionV1 struct {
+	ExtensionPointID string `json:"extensionPointId"`
+	Title            string `json:"title"`
+	Description      string `json:"description"`
+	Type             string `json:"type"`
+}
+
+type ExtensionsV2 struct {
+	AddedLinks        []AddedLink        `json:"addedLinks"`
+	AddedComponents   []AddedComponent   `json:"addedComponents"`
+	ExposedComponents []ExposedComponent `json:"exposedComponents"`
+	ExtensionPoints   []ExtensionPoint   `json:"extensionPoints"`
+}
+
+type Extensions ExtensionsV2
+
+func (e *Extensions) UnmarshalJSON(data []byte) error {
+	var err error;
+	var extensionsV2 ExtensionsV2;
+	
+	if err = json.Unmarshal(data, &extensionsV2); err == nil {		
+		e.AddedComponents = extensionsV2.AddedComponents;
+		e.AddedLinks = extensionsV2.AddedLinks;
+		e.ExposedComponents = extensionsV2.ExposedComponents;
+		e.ExtensionPoints = extensionsV2.ExtensionPoints;
+		
+		return nil;
+	}
+
+	// Fallback (V1)
+	var extensionsV1 []ExtensionV1;
+	if err = json.Unmarshal(data, &extensionsV1); err == nil {
+		// Trying to process old format and add them to `AddedLinks` and `AddedComponents`
+		for _, extensionV1 := range extensionsV1 {
+			if (extensionV1.Type == "link") {
+				extensionV2 := AddedLink{
+					Targets:     []string{extensionV1.ExtensionPointID},
+					Title:       extensionV1.Title,
+					Description: extensionV1.Description,
+				}
+				e.AddedLinks = append(e.AddedLinks, extensionV2);
+			} 
+			
+			if (extensionV1.Type == "component") {
+				extensionV2 := AddedComponent{
+					Targets:     []string{extensionV1.ExtensionPointID},
+					Title:       extensionV1.Title,
+					Description: extensionV1.Description,
+				}
+
+				e.AddedComponents = append(e.AddedComponents, extensionV2);
+			}
+		}
+
+		return nil;
```

Comment:
Correct med if I'm wrong, but this fallback will not take us all the way since the legacy schema doesn't support extension points? So if we silently migrate this, things may still not work as expected in the app. Wouldn't it be better to simply drop support for extensions V1 format and always require the new schema format?

Reply:
No no, you are right that it's not gonna take us all the way.

This backwards compatibility is only in place, so any production deploys of the internal app plugins won't break once they receive this update (we have previously updated their plugin.json to contain the now outdated `extensions: []` meta info. We could see and specify what Grafana versions are supported in cloud, so this is more just to be safe. 

I hope that the following two things will help them to update:
- the plugin.json schema only supports the new version and will throw warnings
- if they introduce an extension point (and have the old plugin.json schema), then in dev mode it won't work for them

Did that answer your question mate?

Reply:
I see, thanks for clarifying!